pkg/endpoint/bpf.go

```diff
@@ -694,26 +672,32 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir, reason string) (uint64, err
 			return 0, fmt.Errorf("unable to regenerate policy for '%s': %s", e.PolicyMap.String(), err)
 		}
 
+		// Synchronously try to update PolicyMap for this endpoint. If any
+		// part of updating the PolicyMap fails, bail out and do not generate
+		// BPF. Unfortunately, this means that the map will be in an inconsistent
+		// state with the current program (if it exists) for this endpoint.
+		// GH-3897 would fix this by creating a new map to do an atomic swap
+		// with the old one.
+		err := e.syncPolicyMap()
+		if err != nil {
+			return 0, fmt.Errorf("unable to regenerate policy because PolicyMap synchronization failed: %s", err)
+		}
+
 		// Walk the L4Policy for ports that require
-		// an L7 redirect and add them to the endpoint; update the L4PolicyMap
-		// with the redirects.
-		c.Mutex.Lock()
-		if c.L4Policy != nil {
-			desiredRedirects, err = e.addNewRedirects(owner, c.L4Policy)
+		// an L7 redirect and add them to the endpoint.
+		if e.DesiredL4Policy != nil {
```

Comment:
By this do you mean making a local copy and then using that instead of the field directly? The `l4Policy` that we assign above is in a separate function?

Reply:
Oh, you are right. I had unrolled the diff to check if `l4Policy` was in the same function, and then misread it. You are also correct in what I had in mind. It does make it a little more indirect, so it might be a bad idea. If we had made that variable in the function anyway, then using it consistently would swing me the other way, since we don't it's a little bit of a toss-up. My motivation is to not accidentally access a different instance of DesiredPolicy midway through a function, or its deferreds and goroutines. In this case we are protected by the lock and we don't use the variable that way, so it's up to you.